Discard, start with zeroes, write and verify in parallel

wjt opened the pull request:
Ready for review.

* [x] Add tests for existing disk-writing code
* [x] Fix bugs exposed by said tests
* [x] Run `blkdiscard` on the whole device
* [x] Write 1MiB of zeros at the start of the device; only write the real data once everything else is written and verified successfully
* [x] Verify and write in parallel
* [x] Decompress in parallel too

Tests & handwave-y performance comparisons:

* [x] xz in vbox
  * ~16% faster than current release, 9% excluding verification step
* [x] gz in vbox
  * ~ 33% faster than current release, 22% excluding verification step
* [ ] uncompressed in vbox
* [x] xz'd ISO in vbox
  * ~ 55% faster; 8% percent faster excluding verification step.
* [ ] gz'd ISO in vbox
* [ ] Test on physical hardware which supports blkdiscard
  * NB: not sure if I have any…

https://phabricator.endlessm.com/T13661

wjt committed:
install: factor out verification code

wjt committed:
install: parametrize keyring path

This makes it easier to write tests.

wjt committed:
install: start a test suite for GisScribe

glibtests.m4 and glib-tap.mk are from GLib Git, with a couple of path
changes in the latter, marked with comments.

This version of tap-test comes from ostree. Unlike the upstream version,
it runs tests in a subdirectory of /var/tmp rather than in the
source/build dir. This is of interest if one is building in a VirtualBox
shared folder, for instance, which doesn't support cool features like
hardlinks. (Without this, the libglnx tests, coming soon, fail.)

wjt committed:
install: don't report progress reading signature

This was a pre-existing bug. It's sort of a hack to use the file size to
distinguish but it does work.

wjt committed:
install: refactor logic to pulse progress bar

wjt committed:
install: move disk writing to GisScribe

The only non-UI logic left in GisInstallPage is getting an FD for the
drive from UDisks. This leaves GisScribe broadly independent of the
surrounding system (with the exception of a harmless call to
"partprobe") and so testable.

wjt committed:
install: more test cases

Before introducing parallel verification, it'd be nice to shake out bugs
in the current implementation, then ensure they don't regress in the
next step.

wjt committed:
install: handle (and test) write errors

I promise this wasn't a regression introduced by this refactoring!

The use of memfd to create an fd where writes past a certain point fail
is due to @pwithnall, and is much simpler than my previous attempt using
a pipe. Thanks!

wjt committed:
install: write zeros at start of disk

wjt committed:
install: blkdiscard target device

Rather than invoking `blkdiscard` the command, which would require
using our favourite pkexec trick to escalate that process, we can just
call the two necessary ioctls.

One could argue that we should only blkdiscard the region we're about to
write, but the data after that region is spiritually overwritten, too,
with undefined contents.

wjt committed:
Import libglnx

This will simplify error handling when calling functions not wrapped by
GIO.

wjt committed:
autogen.sh: remove PKG_NAME

    $ ./autogen.sh
    /usr/bin/gnome-autogen.sh
    ***Warning*** PKG_NAME is deprecated, you may remove it from autogen.sh

pwithnall commented on .gitignore:
```diff
@@ -10,6 +10,8 @@ Makefile
 aclocal.m4
 autom4te.cache
+/build-aux
+compile
```
I normally find it’s easier to use [git.mk](https://github.com/behdad/git.mk) to generate the `.gitignore`, but that’s an FYI rather than a black mark on the review.

wjt replied:
Thanks; I noticed it in `eos-updater`, but not one for this branch.

pwithnall commented on Makefile.am:
```diff
@@ -2,7 +2,14 @@ ACLOCAL_AMFLAGS = -I m4 ${ACLOCAL_FLAGS}
+    po \
+    data \
+    eos-installer-data \
+    tests
```
Nitpick: I find it’s nicer to finish lists in automake files with a `$(NULL)` on its own, to avoid diff noise from adding backslashes in future. I realise this is pedantry.

ebassi replied:
If you don't want to define `NULL =` and clutter your `Makefile.am`, you can always use `$()` to terminate a list

wjt replied:
I've never bothered defining `NULL =`! I used that idiom elsewhere in this branch but not here for some reason. Thanks.

pwithnall commented on autogen.sh:
```diff
@@ -18,5 +16,12 @@ which gnome-autogen.sh || {
+sed -e 's,$(libglnx_srcpath),libglnx,g' < ext/libglnx/Makefile-libglnx.am >ext/libglnx/Makefile-libglnx.am.inc
+
+REQUIRED_AUTOMAKE_VERSION=1.13
 . gnome-autogen.sh
```
Note: gnome-common (and hence `gnome-autogen.sh`) has been deprecated for a long time. [Porting away from it](https://wiki.gnome.org/Projects/GnomeCommon/Migration) should be fairly easy, but it’s obviously a separate issue from this task. And if you plan to port to Meson in the near future, you might want to skip porting away from gnome-common entirely.

wjt replied:
Yep, definitely not one for this branch. All this kind of modernisation is behind "deleting all the unused gnome-initial-setup code" in my priorities

pwithnall commented on configure.ac:
```diff
@@ -54,6 +57,8 @@ PKG_CHECK_MODULES(INITIAL_SETUP,
                   evince-document-3.0
                   webkit2gtk-4.0)
 
+PKG_CHECK_MODULES(LIBGLNX, gio-unix-2.0 >= $GLIB_REQUIRED_VERSION)
```
Nitpick: This is all underquoted, but the practical difference that makes is zero.
```
PKG_CHECK_MODULES([LIBGLNX],[gio-unix-2.0 >= $GLIB_REQUIRED_VERSION])
```

wjt replied:
Tell that to [ostree](https://github.com/ostreedev/ostree/blob/ae61321046ad7f4148a5884c8c6c8b2594ff840e/configure.ac#L105-L109)!

pwithnall commented on ext/Makefile.am:
```diff
@@ -0,0 +1,17 @@
+EXTRA_DIST += libglnx/Makefile-libglnx.am
+noinst_LTLIBRARIES += libglnx.la
+
+# Disable libglnx tests for now
```
Howcome? Should this have a `FIXME` comment and a follow-up task?

wjt replied:
Tedious reasons involving `link` and `renameat` not working on VirtualBox shared folders, which I have now fixed by importing ostree's fork of `tap-test` which runs tests in a tmpdir in `/var/tmp`.

pwithnall commented on gnome-image-installer/Makefile.am:
```diff
@@ -36,6 +36,7 @@ gnome_image_installer_SOURCES = \
 	gnome-image-installer.c gnome-image-installer.h
 
 gnome_image_installer_LDADD =	\
+	$(top_srcdir)/ext/libglnx.la \
```
Not `$(top_builddir)`?

wjt replied:
It should, thanks.

pwithnall commented on tap-test:
```diff
@@ -0,0 +1,5 @@
+
+# run a GTest in tap mode. The test binary is passed as $1
+
+$1 -k --tap
```
You need to add the following to the top-level `Makefile.am` or `distcheck` will not be happy:
```
EXTRA_DIST += glib-tap.mk tap-driver.sh tap-test
```
(with appropriate indentation)

I suggest smoketesting the branch with `make distcheck` and also with `srcdir` ≠ `builddir`.

wjt replied:
Good spot.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+G_DEFINE_TYPE (GisScribe, gis_scribe, G_TYPE_OBJECT);
+
+enum {
+  PROP_0,
```
Nitpick: `PROP_0` can be dropped by setting `PROP_IMAGE = 1`

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+
+G_DEFINE_TYPE (GisScribe, gis_scribe, G_TYPE_OBJECT);
+
+enum {
```
Nitpick: Turn this into a `typedef` and cast the `guint property_id` to that in your `get_property` and `set_property` functions, and instantly profit from `-Wswitch-enum`!

wjt replied:
Ok

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+  guint set_indeterminate_progress_id;
+} GisScribe;
+
+G_DEFINE_QUARK(install-error, gis_install_error);
```
Nitpick: Missing space before `(`

ebassi replied:
Nitpick: drop the trailing `;`: it's not needed.

wjt replied:
Inherited from the existing code, but will fix.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+
+G_DEFINE_QUARK(install-error, gis_install_error);
+
+G_DEFINE_TYPE (GisScribe, gis_scribe, G_TYPE_OBJECT);
```
Nitpick: `G_DEFINE_TYPE` conventionally doesn’t have a `;` after it (same for `G_DEFINE_QUARK` above)

wjt replied:
ok.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+
+  if (self->drive_fd != -1)
+    close (self->drive_fd);
+  self->drive_fd = -1;
```
`gpg_pid`, `gpg_stdout` and `decompressed` don’t seem to be freed here or in `dispose()`. Are they freed elsewhere, unowned, or leaked?

wjt replied:
`gpg_pid` is an integer. `gpg_stdout` is cleared in `gis_scribe_gpg_watch`. I will clear `decompressed` in `dispose`, thanks

pwithnall replied:
Sorry, I should have been clearer about `gpg_pid`: you might want to kill the child process if it’s not `-1`, or perhaps assert if it’s not `-1`? Just a way of preventing code bugs causing leaks.

Similarly for `gpg_stdout`, if you clear it elsewhere, then you can add a `g_assert (self->gpg_stdout == NULL)` to make that more obvious here.

wjt replied:
I'm not going to address this because as explained elsewhere the code that calls GPG will be changed anyway to use GSubprocess.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+  object_class->dispose = gis_scribe_dispose;
+  object_class->finalize = gis_scribe_finalize;
+
+  props[PROP_IMAGE] = g_param_spec_object (
```
Nitpick: Properties are missing gtk-doc comments (I tend to add them, even for internal classes, just to document the code a bit better).

wjt replied:
The documentation is in the blurb for each property

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+
+  /* Providing both the path and the fd seems redundant. However: in the app
+   * proper, we open the fd using udisks so we can write to it as an
+   * unpriveleged user, but also need to be able to run a script on the device,
```
s/unpriveleged/unprivileged/

wjt replied:
Thanks

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+  GFile *signature;
+  gchar *keyring_path;
+  gchar *drive_path;
+  gint drive_fd;
```
This should be in the ‘shared with an modified by the worker thread’ section below.

wjt replied:
Okay. Actually I don't think it's strictly necessary since once the worker thread is spawned it's only accessed on the main thread from `finalize`, by which point the thread is dead.

pwithnall replied:
Then it shouldn’t be accessed from within the mutux. Either is fine, but consistency is good.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+
+  GInputStream *decompressed;
+
+  GMutex mutex;
```
Seems like you could safely use atomics for all three members here which are shared between threads, since none of them are modified in the same critical section. The mutex also works though, so there’s no massive need to change this.

wjt replied:
I did think about it but `bytes_written` is a `guint64` whereas `g_atomic_int_*` operates on pointer-sized `int`. I don't think anyone actually runs `eos-installer` on arm, with 32-bit pointers, but we do build and ship it there.

pwithnall replied:
Good point. As you were.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+  g_mutex_unlock (&self->mutex);
+
+  self->progress = ((gdouble) bytes_written) / ((gdouble) self->image_size);
+  g_object_notify_by_pspec (G_OBJECT (self), props[PROP_PROGRESS]);
```
It would be better to only emit the notification if the value of `self->progress` has actually changed.

pwithnall replied:
(Same for the other property notifications in this branch.)

wjt replied:
Since our smallest image is 7 GiB uncompressed and we write 1 MiB to disk at a time, if the bytes written (and hence progress) does not change at least once per second (the frequency of this callback) then writing the image to disk would take more than 7168 seconds, ie 2 hours. It doesn't, even on a slow spinning disk, so in practice (and notwithstanding double precision floating point accuracy) the value of `self->progress` has almost certainly changed.

pwithnall replied:
Yes, but in the absence of actual bugs in your code, I have to comment on something to make the code review worthwhile, right?

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+                              GCancellable  *cancellable,
+                              GError       **error)
+{
+  gssize buffer_size = 1 * 1024 * 1024;
```
Could be `const`

wjt replied:
Or a `#define`!

pwithnall replied:
¬_¬

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+                              GError       **error)
+{
+  gssize buffer_size = 1 * 1024 * 1024;
+  g_autofree gchar *buffer = g_malloc0 (buffer_size);
```
Does this need to be `g_malloc0()` rather than `g_malloc()`?

ebassi replied:
It depends if `g_input_stream_read()` will pad out with `\0` as the last read byte.

pwithnall replied:
Shouldn’t do: the code only ever writes up to `r` bytes of the buffer, where `r` is the number of bytes read into it.

wjt replied:
It will actually become `if (posix_memalign (&buffer, sysconf(_SC_PAGESIZE), BUFFER_SIZE)) oh_no()` in a future release

wjt replied:
It is currently `g_malloc0` because that's what the old code used.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+gis_scribe_convert_to_mbr (GisScribe *self,
+                           GError   **error)
+{
+  static const char *cmd = "/usr/sbin/eos-repartition-mbr";
```
Does this get any benefit from being `static`?

wjt replied:
Probably not. It was like that when I found (ahem, originally wrote) it.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+  gint fd = -1;
+  g_autoptr(GOutputStream) output = NULL;
+  gboolean ret;
+  GError *error = NULL;
```
Nitpick: While the memory management here is correct for `error`, I think it’s generally clearer to use `g_autoptr(GError) error = NULL` and `g_steal_pointer()` to document the ownership transfers. Note that does change the most likely bug from being a leak into being a double-free (due to omitting a `g_steal_pointer()` rather than a `g_clear_error()`).

wjt replied:
I disagree in this case since `error` must always be passed to `g_task_return_error()` if non-`NULL`, except in one case where it is manually cleared. There is no benefit to littering this function with `g_steal_pointer()`

pwithnall replied:
You say ‘littering’, I say ‘documenting the ownership transfers’. Up to you.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+  self->set_indeterminate_progress_id =
+    g_idle_add_full (G_PRIORITY_DEFAULT_IDLE,
+                     gis_scribe_set_indeterminate_progress,
+                     g_object_ref (self), g_object_unref);
```
Would `g_main_context_invoke()` be more appropriate here? Or do you want the ability to remove the handler from somewhere else in the code?

pwithnall replied:
Seems you do. Ignore me.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+                        GAsyncReadyCallback callback,
+                        gpointer            user_data)
+{
+  GTask *task = g_task_new (self, cancellable, callback, user_data);
```
`task` is leaked on the `self->started` code path. Would be better to use `g_autoptr()` and `g_steal_pointer()` as appropriate.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,820 @@
+  return g_task_propagate_boolean (task, error);
+}
+
+guint
```
Nitpick: Would be helpful to add gtk-doc comments, even though this is internal code. Helps with internal documentation, especially GIR annotations.

wjt replied:
Later -- this branch is not finished, and in particular the concept of there being more than one step will go away.

pwithnall replied:
Doesn’t seem to have happened, unless it happens in a later commit. If so, ignore me. (I’m reviewing the commits sequentially.)

pwithnall commented on tests/Makefile.am:
```diff
@@ -0,0 +1,56 @@
+
+test_scribe_SOURCES = test-scribe.c
+test_scribe_CFLAGS = \
+  $(INITIAL_SETUP_CFLAGS) \
```
Nitpick: Indentation depth differs from other lists in `Makefile.am`s.

pwithnall commented on tests/Makefile.am:
```diff
@@ -0,0 +1,56 @@
+  $(INITIAL_SETUP_CFLAGS) \
+  $(IMAGE_INSTALLER_CFLAGS) \
+  -I $(top_srcdir)/ext/libglnx \
+  -I"$(top_srcdir)/gnome-image-installer/pages/install" \
```
Nitpick: These are `CPPFLAGS`, not `CFLAGS`. In practice, nobody cares.

pwithnall replied:
(Why’s the second one quoted?)

wjt replied:
To inject a bit of variety into proceedings.

pwithnall commented on tests/Makefile.am:
```diff
@@ -0,0 +1,56 @@
+  $(top_builddir)/ext/libglnx.la \
+  $(top_builddir)/gnome-image-installer/pages/install/libgisinstall.la \
+  -llzma \
+  $(WARN_LDFLAGS)
```
Nitpick: `WARN_LDFLAGS` should go in `test_scribe_LDFLAGS`.

pwithnall commented on tests/sign-file:
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python3
```
Missing copyright header.

pwithnall commented on tests/sign-file:
```diff
@@ -0,0 +1,42 @@
+    with tempfile.TemporaryDirectory() as gnupghome:
+        cmd("gpg", "--batch", "--homedir", gnupghome, "--import", a.privkey)
+        cmd("gpg", "--batch", "--homedir", gnupghome, "--detach-sign", "--armor", a.file)
+        cmd("gpgconf", "--homedir", gnupghome, "--kill", "gpg-agent")
```
This is a delightfully small and beautiful script.

pwithnall commented on tests/test-scribe.c:
```diff
@@ -0,0 +1,569 @@
+{
+  gis_scribe_write_async (fixture->scribe, fixture->cancellable,
+                          test_scribe_write_cb, fixture);
+  g_main_loop_run (fixture->loop);
```
Note: When writing unit tests (and in general), I find it’s easier to not use a `GMainLoop` and instead iterate the `GMainContext` manually:
```
static void
async_result_cb (GObject *source_object, GAsyncResult *result, gpointer user_data)
{
  GAsyncResult **result_out = user_data;
  *result_out = g_object_ref (result);
}

static void
test_bad_signature (…)
{
  g_autoptr(GAsyncResult) result = NULL;

  gis_scribe_write_async (fixture->scribe, fixture->cancellable, async_result_cb, &result);

  while (result == NULL)
    g_main_context_iteration (NULL, TRUE);

  ret = gis_scribe_write_finish (scribe, result, &fixture->error);
  …

  g_assert (fixture->finished);
  …
}
```

That means you don’t need to split your test across two functions, and potentially means you can eliminate things like `finished` and `error` from `Fixture`. It also makes the termination conditions for the main loop clearer, which is useful when they are more complex than this.

wjt replied:
This does make things smaller. Thanks!

pwithnall replied:
(I assume you changed this round in a commit later than e0fa060e25f52182bd127dfc134ff0850a2943ee. I’m working my way through the commits sequentially.)

pwithnall reviewed (commented):
I haven’t reviewed the tests or the changes to `gis-install-page.c` in detail. The rest I’ve taken a moderately close look at.

ebassi commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,360 @@
+      return;
+    }
+
+  if (!g_spawn_async_with_pipes (NULL, (gchar **) args, NULL,
```
I do wonder if this wouldn't be better suited by `GSubprocess` and `GOutputStream` for the stdout pipe. It would allow you do avoid using `g_io_channel_unix_new()` and `g_io_add_watch()`, and you'd be able to read the output buffer of the subprocess synchronously inside a `GTask` thread.

wjt replied:
The next step is to pipe the file into gpg so we can verify and write in parallel. I do indeed intend to use `GSubprocess` for that.

wjt replied:
I didn't end up pushing this onto a thread but I may do that if I feel keen tomorrow.

ebassi commented on autogen.sh:
```diff
@@ -4,8 +4,6 @@
 srcdir=`dirname $0`
 test -z "$srcdir" && srcdir=.
 
-PKG_NAME="eos-installer"
```
You could also replace `gnome-autogen.sh` with `autoreconf -v -f -i`.

pwithnall replied:
(cf. https://wiki.gnome.org/Projects/GnomeCommon/Migration)

wjt replied:
Look, I was just annoyed about seeing a warning at the start of every clean build. I have no plans to do any build system migration until after deleting all the unused gnome-initial-setup code, at which point one may as well move to Meson. I can drop this patch if the idea of removing one warning without doing a total rewrite of `autogen.sh` is controversial.

pwithnall replied:
I think what you’ve done here is fine: I was just providing some further information which might be useful if you were planning to drop the gnome-common dependency. Sorry if that wasn’t clear.

wjt commented:
Pushed many fixups. Here's the diff: https://gist.github.com/wjt/e938937829432572111d04a82c549336

Everything not yet addressed is on hold until I finish the features for this PR...

wjt committed:
install: use GSubprocess to launch GPG

A little more verbose, but also a little less cumbersome, than
g_spawn_async_with_pipes().

wjt committed:
install: don't assume PROGRESS message is well-formed

If the current/full values are mis-parsed, that's no huge deal (maybe
the progress bar will behave weirdly) but if there are fewer than 6
fields we would previously crash.

While we're here: log unhandled messages. It would be nice to use the
GOODSIG/BADSIG messages to report in what way the signature is invalid.

wjt committed:
install: verify and write in parallel

wjt committed:
install: pass correct image size to GPG

Historically, GPG has only used the hint when we're installing from
/dev/mapper/endless-image, whose size cannot be determined with stat().
So it hasn't mattered that the "wrong" value is being used.

But when we move to reading the file and tee-ing it to GPG and the
decompressor at the same time, the distinction actually does matter,
because GPG will never be able to determine the size of its input (it's
a pipe) and will always rely on the hint.

The kink here is that when the image is a SquashFS, the compressed
size we store is the size of the SquashFS, but the data we feed to GPG
is from /dev/mapper/endless-image, which corresponds to the uncompressed
image inside the SquashFS. So there needs to be a special-case
somewhere. I chose to put it outside GisScribe because that class knows
nothing whatsoever about SquashFSes.

While we're here: g_file_info_get_size() returns a goffset, which is
signed! We store the uncompressed size as guint64, so let's store the
(always non-negative) compressed size as guint64 too. We already
implicitly assume that it's non-negative since g_format_size_full()
accepts a guint64.

wjt committed:
install: decompress in a subprocess

On Linux, the default pipe buffer size is 64 KiB. Since we are reading
and writing 1 MiB buffers from and to each pipe, it shouldn't be too
surprising that this improves throughput. Without this detail,
decompressing in a subprocess is actually slower than just doing it in
the write thread; but with it, it's faster. (As it happens, 1 MiB is
also the maximum size an unprivileged process can set for pipe buffers.)

A nice way to explore this is to compare the runtimes of these two
commands:

                     xz -cd foo.img.xz | dd bs=1M iflag=fullblock ...
    cat foo.img.xz | xz -cd            | dd bs=1M iflag=fullblock ...

The former is what the most performant of our many shell scripts does,
and on my system with a given ~2GB base image it takes ~3:20. The latter
is closer to what this application is doing, and takes ~3:50. All pipes
here have the default buffer size. (I have inspected the cat, xz and dd
source code to confirm this.)

GisScribe with default pipe buffer sizes takes over 4 minutes (one
run took over 5 minutes, but I think this must be an outlier). With 1
MiB pipe buffers, it's around ~3:35. We can conclude that it's the pipes
into xz and GPG that are important, not the one from xz to the write
thread. This makes some sense since we are reading many more buffers
from xz than we write to it (that's the point of decompression!).

We might be able to do better by interleaving the writes in the tee
thread. Right now it's:

    while buffer = input.read(1024 * 1024):
        gpg.write_all(buffer)
        xz.write_all(buffer)

You could imagine interleaving the writes to gpg and xz, to avoid
starving xz when gpg is not ready to read:

    b = 1024 * 1024
    while buffer = input.read(b):
        x = g = b

        while x or g:
            if g:
                g -= gpg.write(buffer + b - g, g)

            if x:
                x -= xz.write(buffer  + b - x, x)

But let's stick with "straightforward and reasonably performant".

wjt committed:
install: page-align all buffers

This is rumoured to improve throughput.

wjt committed:
install: always write a full buffer to disk

This is rumoured to improve throughput.

wjt committed:
install: check size written against image size

This is vanishingly unlikely to happen in practice given the way images
are constructed, but no harm in checking.

wjt committed:
install: log timings to the journal

This is crude but it's useful when comparing the performance of
different scenarios.

wjt committed:
install: always show warning dialog

Previously, closing the window during the verification step was safe
since nothing was written to disk yet. But that's no longer true: we
start writing immediately, and as soon as it's done advance to the next
page. So as long as this page is visible, the write operation is in
progress.

wjt commented:
This branch is now (I believe) feature-complete!

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,362 @@
+
+  GPid gpg_pid;
+  GIOChannel *gpg_stdout;
+  guint gpg_watch;
```
I still think there should be some assertions checking the state of `gpg_pid`, `gpg_stdout` and `gpg_watch` in `dispose()`/`finalize()` (checking that they have been cleared).

wjt replied:
These variables no longer exist at the top of the branch.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,362 @@
+   * [GNUPG:] PROGRESS /dev/mapper/endless- ? 676 4442 MiB
+   */
+  arr = g_strsplit (line, " ", -1);
+  curr = g_ascii_strtod (arr[4], NULL);
```
It would be good if this code could be a little more robust than simply assuming `arr` is going to have the right number of elements. `gpg` might change its output format at some point.

pwithnall replied:
Should probably also check that the `g_ascii_strtod()` calls don’t fail.

pwithnall replied:
Ah, it seems you add the code to verify the size of `arr` in 870e45b17ba99cae09e014e2ae14e84200248798. Excellent.

I still suggest `g_ascii_strtod()` should get some error checking though.

wjt replied:
I think the failure mode here is quite benign (progress bar doesn't work properly) and unlikely that I find it a bit hard to be too worried. But I have discovered `g_ascii_string_to_unsigned` now exists which means we can both be happy with a mere 6 additional lines of code!

pwithnall replied:
Note that `g_ascii_string_to_[un]signed()` is only for integers, not doubles. There is no handy replacement for `g_ascii_strtod()` with improved error handling (yet).

wjt replied:
Yep. The [specification for this message](https://git.gnupg.org/cgi-bin/gitweb.cgi?p=gnupg.git;a=blob;f=doc/DETAILS;h=0be55f4d;hb=refs/heads/master#l1043) (also linked in the code) doesn't explicitly say whether the numbers are integers, but fortunately I have the GPG source code:

```c
g10/progress.c-123-  snprintf (buffer, sizeof buffer, "%.20s ? %lu %lu %c%s",
g10/progress.c-124-            what? what : "?", current, (unsigned long)total,
g10/progress.c-125-            units[unitidx],
g10/progress.c-126-            unitidx? "iB" : "");
g10/progress.c:127:  write_status_text (STATUS_PROGRESS, buffer);
```

Perhaps I'll send a doc patch.

wjt replied:
https://dev.gnupg.org/T3501

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,362 @@
+  if (g_spawn_check_exit_status (status, NULL))
+    {
+      g_task_return_boolean (task, TRUE);
+      // TODO
```
I assume this is handled in a follow-up commit (I’m working my way through the commits in order).

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,362 @@
+    }
+  else
+    {
+      g_task_return_new_error (task, GIS_INSTALL_ERROR, 0,
```
Why is the error code `0` and not some value from an enum?

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,362 @@
+  if (!g_file_query_exists (self->signature, NULL))
+    {
+      g_task_return_new_error (
+          task, GIS_INSTALL_ERROR, 0,
```
Why is the error code `0` and not some value from an enum?

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,362 @@
+          task, GIS_INSTALL_ERROR, 0,
+          _("The Endless OS signature file \"%s\" does not exist."),
+          signature_path);
+      g_object_unref (task);
```
The refcounting here is correct, but it’s pretty unusual for a function to steal a reference to a `GTask` which comes in as a parameter.

wjt replied:
It no longer does.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -0,0 +1,362 @@
+                                    gis_scribe_gpg_progress,
+                                    task);
+
+  g_child_watch_add (self->gpg_pid, gis_scribe_gpg_watch, task);
```
I’d whack a `g_steal_pointer()` here to make the ownership transfer into the closure for `g_child_watch_add()` more obvious. (It would make no functional difference though: purely documentational.)

wjt replied:
Code no longer exists

pwithnall commented on tests/test-scribe.c:
```diff
@@ -0,0 +1,330 @@
+  fixture->signature = g_file_new_for_path (data->signature_path);
+
+  target_path = g_build_filename (fixture->tmpdir, "target.img", NULL);
+  g_assert (target_path != NULL);
```
Note: GLib has grown `g_assert_nonnull (target_path)`, which you might find nicer to use in future (the error message is slightly more specific than what you get with plain `g_assert()`).

pwithnall replied:
That said, `g_build_filename()` should never fail, so I’m not sure why this assertion is needed.

wjt replied:
I'll drop it.

pwithnall commented on tests/test-scribe.c:
```diff
@@ -0,0 +1,330 @@
+
+  fixture->tmpdir = g_dir_make_tmp ("eos-installer.XXXXXX", &error);
+  g_assert (fixture->tmpdir != NULL);
+  g_assert_no_error (error);
```
Nitpick: I’d put the `g_assert_no_error()` before the assertion about the return value, since otherwise you won’t see the error message on failure — the assertion failure on the return value will end up short-circuiting it.

pwithnall commented on tests/test-scribe.c:
```diff
@@ -0,0 +1,330 @@
+                     const gchar  *basename)
+{
+  gchar *filename = g_test_build_filename (file_type, basename, NULL);
+  g_assert (filename != NULL);
```
`g_test_build_filename()` can’t fail, so this assertion is redundant.

wjt replied:
My mother always told me to assert when things can't fail.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -257,9 +258,18 @@ gis_scribe_gpg_progress (GIOChannel  *source,
+  arr = g_strsplit (g_strchomp (line), " ", -1);
   curr = g_ascii_strtod (arr[4], NULL);
   full = g_ascii_strtod (arr[5], NULL);
+  units = arr[6];
```
Again, I don’t think we should _really_ assume that the format will always have at least as many columns as we want.

pwithnall replied:
(Fixed in commit 870e45b17ba99cae09e014e2ae14e84200248798.)

pwithnall commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -399,168 +216,30 @@ gis_install_page_is_efi_system (GisPage *page)
-    }
-
-  priv->writing = FALSE;
+  g_autofree gchar *msg = g_strdup_printf (_("Step %d of %d"), step, 2);
```
Suggestion: I might add a `g_assert (step <= 2)` here just to make any errors more obvious in future (if more steps are added).

wjt replied:
Ok.

pwithnall commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -588,48 +267,107 @@ gis_install_page_write_cb (GObject      *source,
+
+  if (block == NULL)
+    {
+      g_autoptr(GError) error = g_error_new (GIS_INSTALL_ERROR, 0,
```
Shouldn’t this error code come from an enum?

wjt replied:
Yes, it should, but this pattern of using one domain per page and zero error code is used throughout this application [eg](https://github.com/endlessm/eos-installer/blob/3b3d4229356e5b21e986d7cc85aa7104392f975a/gnome-image-installer/pages/diskimage/gis-diskimage-page.c#L131) [eg](https://github.com/endlessm/eos-installer/blob/3b3d4229356e5b21e986d7cc85aa7104392f975a/gnome-image-installer/pages/disktarget/gis-disktarget-page.c#L367). I do intend to address this in future but I am trying not to expand the scope of this branch to "fix every problem"

pwithnall replied:
Got it. By the 10th time of writing this comment, I was starting to wonder…

pwithnall commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -588,48 +267,107 @@ gis_install_page_write_cb (GObject      *source,
+    {
+      g_autoptr(GError) error = g_error_new (GIS_INSTALL_ERROR, 0,
+                                             _("Internal error"));
+      g_warning ("gis_store_get_object(GIS_STORE_BLOCK_DEVICE) == NULL");
```
Is the `g_warning()` necessary, given that a `GError` is being reported?

wjt replied:
(This is existing code, but...) The `GError` message does not specify what the internal error is. The string `_("Internal error")` is used in a number of "cannot happen" situations in this application. My goal was to report some error in these cases, but not force our translators to translate them.

pwithnall replied:
Explain that, briefly, in a comment in the code? Up to you.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -215,26 +316,240 @@ gis_scribe_class_init (GisScribeClass *klass)
+gis_scribe_convert_to_mbr (GisScribe *self,
+                           GError   **error)
+{
+  const char *cmd = "/usr/sbin/eos-repartition-mbr";
```
Should this be `SBINDIR` or similar?

wjt replied:
If it were, then after `./configure && make` this function would not work because the script is not in `/usr/local/sbin`.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -215,26 +316,240 @@ gis_scribe_class_init (GisScribeClass *klass)
+
+  g_source_remove (timer_id);
+
+  if (r < 0 || error != NULL)
```
Is it ever possible for `(r < 0 && error == NULL)` to be true? (I can’t see how it can be.)

If so, `g_task_return_error()` will be called with a `NULL` error.

If not, the condition could be simplified to `(error != NULL)`.

wjt replied:
This condition is removed later.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -215,26 +316,240 @@ gis_scribe_class_init (GisScribeClass *klass)
+
+  syncfs (self->drive_fd);
+  close (self->drive_fd);
+  self->drive_fd = -1;
```
`drive_fd` is accessed outside the `mutex` here.

wjt replied:
Code no longer exists

pwithnall replied:
Ideally these issues with the non-`HEAD` comments would all be fixed, to save problems with reverts or bisections in future. However, that’s an effort/reward tradeoff decision for you to make, not me.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -215,26 +316,240 @@ gis_scribe_class_init (GisScribeClass *klass)
+    {
+      g_autofree gchar *parse_name = g_file_get_parse_name (self->image);
+      g_task_return_new_error (task,
+                               GIS_INSTALL_ERROR, 0,
```
Error code is not an enum value here.

(I realise all these comments will get the same answer, but I should point out each location where it occurs.)

pwithnall replied:
Same below.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -215,26 +316,240 @@ gis_scribe_class_init (GisScribeClass *klass)
+  g_autofree gchar *basename = NULL;
+  g_autoptr(GConverter) decompressor = NULL;
+  GInputStream *input = NULL;
+  GError *error = NULL;
```
`g_autoptr()`

wjt replied:
I've reviewed every bare `GError *error` in `gis-scribe.c` and I don't think any of them would be made more legible by the introduction of `g_autoptr(GError)` + `g_steal_pointer`.

wjt replied:
I  have relented and changed them all anyway

pwithnall replied:
Excellent. Welcome to the cult.

(More seriously, I definitely appreciate the documentation value of consistent use of `g_autoptr()`, indicating that a variable is owned.)

pwithnall commented on tests/test-scribe.c:
```diff
@@ -222,12 +258,28 @@ static void
+  g_autofree gchar *target_contents = NULL;
+  gsize target_length = 0;
+  g_autofree gchar *expected_contents = g_malloc (IMAGE_SIZE_BYTES);
+  GError *error = NULL;
```
`g_autoptr()`

pwithnall commented on tests/test-scribe.c:
```diff
@@ -222,12 +258,28 @@ static void
+                             &target_contents, &target_length,
+                             &error);
+  g_assert_no_error (error);
+  g_assert (ret);
```
Nitpick: GLib now has `g_assert_true()` which clarifies the types here a little.

pwithnall commented on tests/test-scribe.c:
```diff
@@ -222,63 +231,59 @@ test_scribe_write_cb (GObject      *source,
+  GAsyncResult **result_out = data;
 
-  g_main_loop_quit (fixture->loop);
+  *result_out = g_object_ref (result);
```
Ideally this refactoring to drop the `GMainLoop` would be squashed into the commit which introduced this test code. Up to you.

wjt replied:
After losing an hour of work last week in a failed rebase trying to squash a different change back, I decided not to do this.

pwithnall commented on tests/test-scribe.c:
```diff
@@ -341,23 +347,27 @@ main (int argc, char *argv[])
       .image_path = image_path,
       .signature_path = image_gz_sig_path,
+      .error_domain = GIS_INSTALL_ERROR,
+      /* TODO: .error_code */
```
Is this error code handled in a later commit? (If it is, ignore this comment. I’m working through the commits sequentially.)

pwithnall replied:
Similarly for the other `TODO` comments below.

wjt replied:
No, this is a todo for the `0` error code issue discussed above.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -457,9 +491,18 @@ gis_scribe_write_thread (GTask        *task,
+  if (syncfs (fd) < 0)
+    {
+      glnx_throw_errno_prefix (&error, "syncfs failed");
+      g_task_return_error (task, error);
```
The output stream isn’t closed on this error path (or the one above). Do you trust that you hold the last reference to `output` and that it will be implicitly closed on finalisation?

wjt replied:
It's not closed on the other error paths either. Yes, I do, because it is defined within `gis_scribe_write_thread` and doesn't escape.

pwithnall commented on configure.ac:
```diff
@@ -1,12 +1,12 @@
 AC_REQUIRE_AUX_FILE([tap-driver.sh])
 
-AM_INIT_AUTOMAKE([dist-xz no-dist-gzip foreign tar-ustar])
+AM_INIT_AUTOMAKE([1.13 dist-xz no-dist-gzip foreign tar-ustar subdir-objects])
```
These changes aren’t strictly needed for importing `libglnx`. I’d factor them out into a separate commit. (Up to you; depends whether you think the rebase faff is worth it.)

wjt replied:
I do not!

pwithnall commented on tests/test-scribe.c:
```diff
@@ -312,6 +327,30 @@ test_error (Fixture       *fixture,
+      g_assert_cmpmem (first_mib, read_size,
+                       expected_first_mib, size);
+    }
+  /* else: TODO: check first mib of file on disk */
```
Is this TODO handled in a later commit? (If so, ignore this comment. I’m working my way through the commits sequentially.)

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -366,6 +370,23 @@ gis_scribe_update_progress (gpointer data)
+  range[0] = 0;
+
+  if (ioctl (fd, BLKGETSIZE64, &range[1]))
+    return glnx_throw_errno_prefix (error, "can't get size for blkdiscard");
```
What if this returns `range[1] == 0`? Will `BLKDISCARD` throw a hissy fit?

wjt replied:
I haven't tested but I assume it'll be fine, or at worst fail harmlessly. Looking at [the source for `blkdiscard(8)`](https://git.kernel.org/pub/scm/utils/util-linux/util-linux.git/tree/sys-utils/blkdiscard.c) it doesn't check for a zero-byte block device.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -506,6 +527,13 @@ gis_scribe_write_thread (GTask        *task,
 
   g_thread_yield ();
 
+  if (!gis_scribe_blkdiscard (fd, &error))
```
What is the benefit to running `blkdiscard` if you’re only going to write over that area immediately afterwards? I assume it gives some kind of speed up — but how?

wjt replied:
Quoth Daniel Drake:
> This tells flash-based disks that all data is now unused, which frees up wear levelling algorithms and boosts performance for next time each eraseblock is written. Not all storage devices support blkdiscard so any failure here should be logged but not presented to the user.

I will lift the first sentence into a comment.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -721,27 +730,27 @@ gis_scribe_gpg_progress (GIOChannel  *source,
     }
   else
     {
+      /* TODO: surface more details about the error */
```
Is this TODO fixed in a following commit? (If so, ignore this comment. I’m reviewing commits sequentially.)

wjt replied:
No, it's not.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -764,7 +773,10 @@ gis_scribe_begin_verify (GisScribe *self,
-  gint outfd;
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  g_autoptr(GSubprocess) subprocess = NULL;
+  GInputStream *gpg_stdout = NULL;
```
Nitpick: Don’t assign `NULL` to `gpg_stdout` on initialisation, or it makes it look like it should own a reference to the stream (as if you were to call `g_clear_pointer()` on it later).

wjt replied:
I think the presence of `g_autoptr` on the variables which do own a reference is a good suggestion that this one does not. It's a moot point because the variable is not referenced before it is assigned to so the compiler will probably drop the `= NULL`s…

pwithnall replied:
> I think the presence of `g_autoptr` on the variables which do own a reference is a good suggestion that this one does not.

Yes, except there are functions where you’ve used `GError *error = NULL` without `g_autoptr()` (unless that’s a convention you changed in the latter half of the branch?).

> It's a moot point because the variable is not referenced before it is assigned to so the compiler will probably drop the `= NULL`s…

(I don’t care what the compiler does! I care about the code readability here.)

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -406,6 +406,24 @@ gis_scribe_write_thread_copy (GisScribe     *self,
+
+  if (bytes_written != self->image_size)
+    {
+      g_set_error (error, GIS_INSTALL_ERROR, 0,
```
Error code is not from an enum here.

wjt replied:
As above.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -34,6 +36,28 @@
+typedef enum {
+  GIS_SCRIBE_TASK_TEE        = 1 << 0,
+  GIS_SCRIBE_TASK_VERIFY     = 1 << 1,
+  GIS_SCRIBE_TASK_WRITE      = 1 << 3,
```
Shouldn’t this be `1 << 2`?

wjt replied:
A future commit introduces the `1 << 2` value. I want the order to match the flow of data, and I didn't see the value in defining `WRITE = 1 << 2` and then changing it in the next patch.

pwithnall replied:
So I noticed, later on. The downside of reviewing patches sequentially.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -34,6 +36,28 @@
+static const gchar *
+gis_scribe_task_get_label (GisScribeTask flag)
+{
+  switch (flag)
```
I guess this isn’t ever meant to deal with combinations of the flags?

wjt replied:
It's not. I've tried to follow the convention of using `GisScribeTask` to refer to individual flags and `int` for combinations of.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -232,7 +294,9 @@ gis_scribe_finalize (GObject *object)
   g_clear_pointer (&self->drive_path, g_free);
+  g_clear_error (&self->error);
   g_mutex_clear (&self->mutex);
+  g_cond_clear (&self->cond);
```
Could additionally assert that `outstanding_tasks == 0`.

wjt replied:
Good idea.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -395,6 +478,31 @@ gis_scribe_blkdiscard (gint     fd,
+gis_scribe_write_thread_await_gpg (GisScribe *self,
+                                   GError   **error)
+{
+  GError *_error = NULL;
```
Suggestion: I normally call this variable `local_error` (to distinguish it from the `GError **error` argument).

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -395,6 +478,31 @@ gis_scribe_blkdiscard (gint     fd,
+
+  g_mutex_lock (&self->mutex);
+  /* Wait until all other tasks have ended */
+  while (self->outstanding_tasks != GIS_SCRIBE_TASK_WRITE)
```
Is it ever possible for `self->outstanding_tasks == 0` to become true while in this loop?

wjt replied:
No: this function is ultimately called from `gis_scribe_write_thread` which is the `thread_func` for the `WRITE` task. I guess this check could be:

```c
while ((self->outstanding_tasks & ~GIS_SCRIBE_TASK_WRITE) != 0)
    wait()

assert (self->outstanding_tasks == GIS_SCRIBE_TASK_WRITE)
```

to make that expliit.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -785,46 +911,225 @@ gis_scribe_begin_verify (GisScribe *self,
+  g_slice_free (GisScribeTeeData, data);
+}
+
+/* Reads the image to disk and writes it to both the GPG subprocess and the
```
‘Reads the image *from* disk’?

wjt replied:
oops

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -785,46 +911,225 @@ gis_scribe_begin_verify (GisScribe *self,
+                       GCancellable     *cancellable)
+{
+  g_autofree gchar *buffer = g_malloc0 (BUFFER_SIZE);
+  GError *error = NULL;
```
`g_autoptr()`

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -785,46 +911,225 @@ gis_scribe_begin_verify (GisScribe *self,
+  else
+    {
+      g_debug ("%s: discarding subsequent error from %s: %s", G_STRFUNC,
+               inner_task_name, error->message);
```
I’d add a `g_clear_error (&error);` here, just in case the code below gets refactored later and accidentally sets another error over the top of it. (I don’t think that can currently happen though.)

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -834,16 +1139,52 @@ gis_scribe_write_async (GisScribe          *self,
+  g_autoptr(GInputStream) input = NULL;
+  g_autoptr(GOutputStream) write_pipe = NULL;
+  g_autoptr(GOutputStream) gpg_stdin = NULL;
+  GError *error = NULL;
```
`g_autoptr()`

pwithnall commented on tests/test-scribe.c:
```diff
@@ -310,6 +311,40 @@ test_scribe_write_cb (GObject      *source,
+  g_assert (ret);
+
+  g_assert_cmpmem (expected_contents, ONE_MIB,
+                   target_contents, MIN (ONE_MIB, target_length));
```
Nitpick: Rather than `MIN`, you might want a `g_assert_cmpuint (target_length, <=, ONE_MIB)`.

wjt replied:
That assertion would always fail because `target_length` should be the full length of the image. I guess `target_length, >=, ONE_MIB` would do.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -41,6 +41,7 @@
 typedef enum {
   GIS_SCRIBE_TASK_TEE        = 1 << 0,
   GIS_SCRIBE_TASK_VERIFY     = 1 << 1,
+  GIS_SCRIBE_TASK_DECOMPRESS = 1 << 2,
```
Ah, this explains the `1 << 3` in a previous commit. Ignore my other comment.

wjt replied:
:)

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1098,6 +1066,120 @@ gis_scribe_begin_tee (GisScribe          *self,
+  const gchar *args[] = { NULL, "-cd", NULL };
+  g_autoptr(GSubprocessLauncher) launcher = NULL;
+  GSubprocess *subprocess = NULL;
+  GError *error = NULL;
```
`g_autoptr()`

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1098,6 +1066,120 @@ gis_scribe_begin_tee (GisScribe          *self,
+    {
+      g_critical ("g_file_get_basename returned NULL");
+      g_task_return_new_error (task,
+                               GIS_INSTALL_ERROR, 0,
```
Error code is not from an enum.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1098,6 +1066,120 @@ gis_scribe_begin_tee (GisScribe          *self,
+    }
+  else
+    {
+      g_task_return_new_error (task, GIS_INSTALL_ERROR, 0,
```
Error code is not from an enum.

pwithnall commented on tests/test-scribe.c:
```diff
@@ -542,8 +542,9 @@ main (int argc, char *argv[])
       .signature_path = trunc_gz_sig_path,
-      .error_domain = G_IO_ERROR,
-      .error_code = G_IO_ERROR_PARTIAL_INPUT,
+      /* TODO: report a kinder error than "child process exited with code 1" */
```
TODO comment here. (Ignore this comment if it’s addressed in a later commit. I’m reviewing the commits sequentially.)

pwithnall replied:
And another below.

wjt replied:
I'd expect to fix this at the same time as the pervasive use of `0` error codes. That said, I guess it is worth fixing this because, when the image file is truncated, there's a race between GPG and the decompressor to find out who will notice the error first…

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -502,22 +504,40 @@ gis_scribe_write_thread_await_gpg (GisScribe *self,
 
+/* Allocates size bytes, page-aligned. */
+     G_GNUC_MALLOC
+     G_GNUC_ALLOC_SIZE(1)
```
Odd indentation on these two decorators. Is that intentional?

pwithnall replied:
You haven’t page-aligned the decorators, have you? 🤔

pwithnall commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -285,8 +285,11 @@ gis_install_page_open_for_restore_cb (GObject      *source,
   g_autoptr(GFile) signature = NULL;
   g_autoptr(GisScribe) scribe = NULL;
+  guint64 uncompressed_size = gis_store_get_required_size ();
+  guint64 compressed_size = gis_store_get_image_size ();
```
Nitpick: I’d put `_bytes` as a suffix on these variables names to make it clearer what units we’re working in.

wjt replied:
OK.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -68,6 +68,11 @@ typedef struct _GisScribe {
+   * indicate its progress, since it reads the image data from a pipe. Equal to
+   * 'image_size' if 'image' is uncompressed.
+   */
+  guint64 compressed_size;
```
NItpick: + `_bytes`

wjt replied:
Ok, renamed this and `image_size`.

pwithnall commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -407,13 +428,15 @@ gis_scribe_init (GisScribe *self)
 {
   g_return_val_if_fail (G_IS_FILE (image), NULL);
   g_return_val_if_fail (image_size > 0, NULL);
+  g_return_val_if_fail (compressed_size > 0, NULL);
```
`0` is allowed as a valid value in the `GParamSpec`. Which one is correct?

wjt replied:
A GPT disk image must be at least 1 + (2 * 33) 512-byte sectors long (MBR, two copies of (GPT header plus at least 16384 bytes for partition entries)). You could imagine a compression format that compresses a degenerate GPT disk image to one byte. I'll add these minima.

pwithnall reviewed (commented):
Whew. Looks generally good to me. Some comments (many of them repeats of the same questions, so a lot of them could be addressed at the same time).

wjt committed:
install: add _bytes suffix to variables

wjt committed:
install: set more accurate minima for image sizes

wjt committed:
install: parse GPG progress as uint64, with error handling

wjt committed:
install: GError * -> g_autoptr(GError)

wjt committed:
install: add a few scraps of documentation

wjt committed:
install: document _("Internal error")

wjt commented:
Pushed some patches to address many bits of review feedback. I've left those that I will squash  before merge unsquashed for ease of reviewing.

pwithnall commented on tests/test-scribe.c:
```diff
@@ -359,8 +359,9 @@ assert_first_mib_zeroed (Fixture *fixture)
   g_assert_no_error (error);
   g_assert (ret);
 
+  g_assert_cmpuint (target_length, >=, ONE_MIB);
```
`<=`, no?

wjt replied:
No, the file on disk or memfd is always larger than 1 MiB.

wjt replied:
Pushed one more patch.

pwithnall replied:
Hmm, I was going by the `g_input_stream_read_all()` call in the memfd case, where you definitely wouldn’t expect `target_length > ONE_MIB`, since that would mean it had read off the end of the buffer it was provided. But asserting `target_length >= ONE_MIB` does work for both cases if you ignore the possibility of reading off the end of the buffer, and want to catch `target_length < ONE_MIB` as an error (which makes sense).

👍

wjt replied:
I guess one could make this clearer by floating the two slightly different assertions into each side of the `if`. I'll take the liberty of doing that then squashing & merging.

pwithnall commented:
> Pushed some patches to address many bits of review feedback. I've left those that I will squash before merge unsquashed for ease of reviewing.

One minor comment; otherwise all the fixups look good to me.

pwithnall reviewed (approved):
All aboard the squash-and-merge bus, as far as I’m concerned.
